src/client/js/components/circle-chart.js

```diff
@@ -0,0 +1,244 @@
+    if (newValue === 'undefined' || newValue === oldValue) {
+      return
+    }
+
+    switch (name) {
+      case 'data':
+        this.data = JSON.parse(newValue)
+        this.createOrUpdateChart()
+        break
+      case 'is-donut':
+        this.isDonut = JSON.parse(newValue)
+        break
+      case 'show-percentage-for':
+        this.showPercentageFor = newValue
+        break
+      case 'title':
+        this.title = newValue
+        break
+      default:
+        console.warning(`Unhandled attribute: ${name}`)
+    }
+  }
+
+  composeCircles () {
+    let sliceOffset = 0
+    return `
+      ${this.data.reduce((acc, curr) => {
+        const percentage = calcPercentage(this.total, curr.count)
+        const innerRadius = this.isDonut ? 0.85 : 0
+        const strokeLength = CHART_CIRCUMFERENCE * percentage
+
+        const circle = `
+          <circle
+            stroke='${curr.color}'
+            stroke-dasharray='${strokeLength} ${CHART_CIRCUMFERENCE}'
+            stroke-dashoffset='${-1 * CHART_CIRCUMFERENCE * sliceOffset}'
+            stroke-width='${CHART_DIAMETER * (1 - innerRadius)}%'
+          ></circle>
+        `
+        acc.push(circle)
+        sliceOffset += percentage
+
+        return acc
+      }, []).join('')}
+    `
+  }
+
+  createChartLabels () {
+    return `
+      ${this.title !== ''
+        ? `<strong class='circle-chart-title'>${this.title}</strong>`
+        : ''}
+      ${this.data.map(({ name, color }) => (
+        `<label class='circle-chart-label' style='color: ${color}'>${name}</label>`
+      )).join('')}
+    `
+  }
+
+  createCircleLabel () {
+    const relevantItem = this.data.find(d => d.key === this.showPercentageFor)
+    if (!relevantItem) {
+      return ''
+    }
+    const percentage = calcPercentage(this.total, relevantItem.count)
+    return `
+      <text
+        dy='${CHART_RADIUS * 0.15}'
+        fill='${relevantItem.color}'
+        font-size='${CHART_RADIUS * 0.4}'
+        font-size='50'
+        x='${CHART_RADIUS}'
+        y='${CHART_RADIUS}'
+      >
+          ${Math.round(percentage * 100)}%
+      </text>
+    `
+  }
+
+  createChart () {
+    // Create SVG with circles
+    const circles = this.composeCircles()
+    this.svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg')
+    this.svg.setAttribute('viewBox', `0 0 ${CHART_DIAMETER} ${CHART_DIAMETER}`)
+    this.svg.innerHTML = `
+      ${circles}
+      ${this.createCircleLabel()}
+    `
+
+    // Create labels
+    this.labels = document.createElement('figcaption')
+    this.labels.innerHTML = this.createChartLabels()
+
+    // Add chart elements to DOM
+    this.chartElement.append(this.svg)
+    this.chartElement.append(this.labels)
+  }
+
+  updateChart () {
+    this.svg.innerHTML = `
+      ${this.composeCircles()}
+      ${this.createCircleLabel()}
+    `
+    this.labels.innerHTML = this.createChartLabels()
+  }
+
+  createOrUpdateChart () {
```

Comment:
The naming of this function confused me at first – is this just an "init" function?  Would the built-in lifecycle callback `connectedCallback` be more appropriate?  Perhaps it might also negate the need for timeout discussed below?

Reply:
This method handles the logic around when to init the circles or when they only need to be updated. When the chart is only being updated and not created from scratch on `data` change it’s being called from here](https://github.com/mozilla/blurts-server/pull/2764/files#diff-c6fd04d9ec6e829967576bae19b241f7f063a5e1d10bf3537675c0cb15d53027R119). Moving it to `connectedCallback` would only render the chart once and not on data updates. Like the example update that currently occurs when clicking on one of the charts.

In the event of an update, the chart is not being recreated from scratch. That would make adding transitions for the chart parts possible in the future. For now, I only opted for a simple fade-in/out which is why I added it. Without the timeout, we would not be able to have a CSS transition. Happy to discuss other approaches as we have the opportunity to make it better when we know how it behaves for our use cases.